firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLitePersistence.java

```diff
@@ -455,6 +459,122 @@ private Cursor startQuery() {
+
+    private int queriesPerformed = 0;
+    private Iterator<Object> argsIter;
+
+    private static final int LIMIT = 900;
+
+    /**
+     * Creates a new {@code LongQuery} with parameters that describe a template for creating each
+     * subquery.
+     *
+     * <p>Each subquery will have the following form:
+     *
+     * <p>[head][an auto-generated comma-separated list of '?' placeholders][tail]
+     *
+     * <p>For example, imagine for demonstration purposes that the limit were 2, and the {@code
+     * LongQuery} was created like this:
+     *
+     * <pre class="code">
+     *     String[] args = {"foo", "bar", "baz", "spam", "eggs"};
+     *     LongQuery longQuery = new LongQuery(
+     *         db,
+     *         "SELECT name WHERE id in (",
+     *         Arrays.asList(args),
+     *         ")"
+     *     );
+     * </pre>
+     *
+     * <p>Assuming limit of 2, this query will issue three subqueries:
+     *
+     * <pre class="code">
+     *     query.performNextPart(); // "SELECT name WHERE id in (?, ?)", binding "foo" and "bar"
+     *     query.performNextPart(); // "SELECT name WHERE id in (?, ?)", binding "baz" and "spam"
+     *     query.performNextPart(); // "SELECT name WHERE id in (?)", binding "eggs"
+     * </pre>
+     *
+     * @param db The database on which to execute the query.
+     * @param head The non-changing beginning of the query; each subquery will begin with this.
+     * @param allArgs The list of host parameters to bind. If the list size exceeds the limit,
+     *     several subqueries will be issued, and the correct number of placeholders will be
+     *     generated for each subquery.
+     * @param tail The non-changing end of the query; each subquery will end with this.
+     */
+    LongQuery(SQLitePersistence db, String head, List<Object> allArgs, String tail) {
+      this.db = db;
+      this.head = head;
+      this.argsHead = Collections.emptyList();
+      this.allArgs = allArgs;
+      this.tail = tail;
+
+      argsIter = allArgs.iterator();
+    }
+
+    /**
+     * The longer version of the constructor additionally takes {@code argsHead} parameter that
+     * contains parameters that will be reissued in each subquery, i.e. subqueries take the form:
+     *
+     * <p>[head][argsHead][an auto-generated comma-separated list of '?' placeholders][tail]
+     */
+    LongQuery(
+        SQLitePersistence db,
+        String head,
+        List<Object> argsHead,
+        List<Object> allArgs,
+        String tail) {
+      this.db = db;
+      this.head = head;
+      this.argsHead = argsHead;
+      this.allArgs = allArgs;
+      this.tail = tail;
+
+      argsIter = allArgs.iterator();
+    }
+
+    /** Whether {@link performNextPart} can be called. */
+    boolean hasMore() {
+      return argsIter.hasNext();
+    }
+
+    /** Performs the next subquery and returns a {@link Query} object for method chaining. */
+    Query performNextPart() {
```

Comment:
I might try to align the naming of hasMore / performNextPart / getQueriesPerformed...

E.g.: hasMoreQueries / performNextQuery / getQueriesPerformed

You could also s/Query/SubQuery/ to create a LongQuery => many SubQueries conceptual relationship... but I don't feel strongly either way.